Added Sign In page

Slartibartfass2 opened the pull request:
Closes #1 

## What I have made

- Added Sign In Page using the input elements designed in #112 
  - the input fields don't use any schemas because all they need to do is matching the credentials on the server
- Added Sign In and Sign Out logic (this was a pain in my ass)
  - the backend endpoint sends back a user and a token, we only use the token because the user is stored in the tokens' payload
  - the token is stored in the local storage and used for every http request that follows
  - when the sign-out button is pressed, the token is removed from the local storage
  - so far so good, now the painful part
  - initially, I tried to decode the jwt in the root `layout.ts`, to that the user is easily accessible on any page
  - **but** svelte loads the `page.ts` and `layout.ts` not always on the client but initially on the web server, where the local storage is obviously not accessible
  - even though it also loads these pages in the client, that's not always the case (afaik only when the user loads a page in the browser, but not if they navigate between different pages of the same app)
  - this really sucked, so now we access the local storage on each page separately
  - note that the call to `getCurrentUser` is also sometimes executed on the server but always another time on the client
  - in between the response is undefined, and the navigation bars handle this case as seen below
  
![2024-12-14 13_14_47-SnowballR – Mozilla Firefox](https://github.com/user-attachments/assets/3f1b0187-bd7a-4f38-a290-6db454d5e42c)

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - ~~[ ] unit tests~~ (I tried mocking the local storage, but somehow I couldn't mock the headers)
  - [x] integration tests
  - ~~[ ] end-to-end tests~~ (not yet available)
- [x] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 committed:
feat: add sign in page

Slartibartfass2 committed:
test: add test case for forgot password link

Slartibartfass2 committed:
feat: add redirect from project/[projectId] to project/[projectId]/dashboard

Slartibartfass2 committed:
feat: add sign in and sign out logic

Merseleo commented on src/lib/components/composites/navigation-bar/UserMenu.svelte:
```diff
@@ -47,6 +48,11 @@
+
+    function signOut() {
+        localStorage.removeItem("token");
+        goto("/signin");
```
I thought we want to explicitly logout? So you need to call the logout method from the BackendController.

Slartibartfass2 replied:
Yes, you're right. I forgot that when I implemented this back then.

I changed it now.

Merseleo commented on src/lib/controller/backend-controller.ts:
```diff
@@ -82,8 +91,13 @@ export class BackendController implements IBackendController {
-        return this.user(0);
+        const jwt = localStorage.getItem("token");
+        if (!jwt) {
+            throw new Error("No JWT stored in the local storage");
```
The reason for this error is, that no user is logged in, but is requested, right? Maybe you can add the reason to error message?

Slartibartfass2 replied:
I replaced this with the `getCurrentUser` method, so this is outdated.

Merseleo commented on src/lib/model/backend.ts:
```diff
@@ -97,3 +97,7 @@ export interface Author {
 export type AuthorSpec = Omit<Author, "id">;
+
+export interface SignInResponse {
+    token: string;
```
If I understood it correctly, you need a token and a refresh token, right? So I guess you need to store both and define both in the interface.

Slartibartfass2 replied:
I'll wait for the new API for this.

Merseleo commented on tests/integration/input/password-input.test.ts:
```diff
@@ -94,4 +100,16 @@ describe("PasswordInput", () => {
         });
     });
+
+    test("When `showForgotPasswordLink` is false, then link is not shown", () => {
```
```suggestion
    test("When `showForgotPasswordLink` is false, then the link is not shown", () => {
```

Slartibartfass2 replied:
Added it

Merseleo commented on src/routes/(auth)/signin/+page.svelte:
```diff
@@ -1,4 +1,86 @@
+            <Alert.Root variant="destructive">
+                <CircleAlert class="size-4" />
+                <Alert.Title>Error</Alert.Title>
+                <Alert.Description>{error}</Alert.Description>
```
```suggestion
                <Alert.Title>{error}</Alert.Title>
```

What do you think of only providing the error and not showing: "Error: ...", as it is red and the error message already indicates that it is an error?

Slartibartfass2 replied:
I agree that's enough.

Merseleo commented on src/routes/(auth)/signin/+page.svelte:
```diff
@@ -1,4 +1,86 @@
+                type="email"
+                bind:value={email}
+            />
+            <PasswordInput class="w-full" showForgotPasswordLink bind:value={password} />
```
Imo the we should not use the PasswordInput component directly here, as it shows what is missing (so at least two numbers etc.), but is this really necessary / usual to show on the login page?

Slartibartfass2 replied:
No, it isn't this was introduced due to the changes in the sign-up PR and rebasing of this branch. I added a flag so that these inputs aren't validated, as they shouldn't.

Merseleo reviewed (commented):
There are only a few small things, we can discuss. To properly test the "Login" we need a running and correct working backend, so let's a bit and hopefully we will get this soon, to test the "Login".

Slartibartfass2 committed:
feat: add logout logic

Slartibartfass2 committed:
feat: add example user routes without SSR

Slartibartfass2 committed:
feat: deactivate validation on sign in page

Slartibartfass2 committed:
feat: add loading and error feedback to sign up page

This was already added for the sign in page and not added to the sign up page earlier.
The loading and error feedback provides a better experience for the user as they can see what is happening.

Slartibartfass2 reviewed (commented):
Thank you for the review

luca-schlecker commented:
Closing in favor of #373
